core/state/snapshot/snapshot.go

```diff
@@ -539,58 +536,23 @@ func diffToDisk(bottom *diffLayer) *diskLayer {
 	base.stale = true
 	base.lock.Unlock()
 
-	// Destroy all the destructed accounts from the database
-	for hash := range bottom.destructSet {
-		// Skip any account not covered yet by the snapshot
-		if base.genMarker != nil && bytes.Compare(hash[:], base.genMarker) > 0 {
-			continue
-		}
-		// Remove all storage slots
-		rawdb.DeleteAccountSnapshot(batch, hash)
-		base.cache.Set(hash[:], nil)
-
-		it := rawdb.IterateStorageSnapshots(base.diskdb, hash)
-		for it.Next() {
-			key := it.Key()
-			batch.Delete(key)
-			base.cache.Del(key[1:])
-			snapshotFlushStorageItemMeter.Mark(1)
-
-			// Ensure we don't delete too much data blindly (contract can be
-			// huge). It's ok to flush, the root will go missing in case of a
-			// crash and we'll detect and regenerate the snapshot.
-			if batch.ValueSize() > 64*1024*1024 {
-				if err := batch.Write(); err != nil {
-					log.Crit("Failed to write storage deletions", "err", err)
-				}
-				batch.Reset()
-			}
-		}
-		it.Release()
-	}
 	// Push all updated accounts into the database
 	for hash, data := range bottom.accountData {
 		// Skip any account not covered yet by the snapshot
 		if base.genMarker != nil && bytes.Compare(hash[:], base.genMarker) > 0 {
 			continue
 		}
 		// Push the account to disk
-		rawdb.WriteAccountSnapshot(batch, hash, data)
-		base.cache.Set(hash[:], data)
-		snapshotCleanAccountWriteMeter.Mark(int64(len(data)))
-
+		if len(data) != 0 {
+			rawdb.WriteAccountSnapshot(batch, hash, data)
+			base.cache.Set(hash[:], data)
+			snapshotCleanAccountWriteMeter.Mark(int64(len(data)))
+		} else {
+			rawdb.DeleteAccountSnapshot(batch, hash)
+			base.cache.Set(hash[:], nil)
+		}
 		snapshotFlushAccountItemMeter.Mark(1)
 		snapshotFlushAccountSizeMeter.Mark(int64(len(data)))
-
```

Comment:
Removing this means we don't flush the batches if they exceed 64k, we only flush at the end. Is this something that we don't need anymore?

Reply:
Yes, we don't need this batching flush anymore for Mainnet and live testnets, reasons:

Before cancun, we have the self-destruction and the deleted account may have a gigantic storage. If we want to wipe them all in a single batch, then the OOM is expected; 
The self-destruction has been deprecated since the cancun fork and we know the largest storage wiping over the past decade (mainnet) is around 15 megabytes, so we can confidently remove this protection mechanism.

And in another pull request, I restore the original behavior back. The reason is that the Geth-based fork chains (e.g. L2s) may or may not still have the self-destruction. And they probably want to keep updating their codebase to align with the latest geth codebase. It would be nice to still have this protection mechanism for them.

However, the flat states will ultimately be moved into pathdb, in which the state flushing (trie nodes and the flat states) MUST be flushed in the single batch. At that point, we must drop this protection mechanism, otherwise state corruption could occur if panic happens in the middle of two batches.